### PR TITLE
importFixes: Support missing "React" at a JSXOpeningElement

### LINKED ISSUE #15456: TypeError: Cannot read property 'flags' of undefined
```ts
import {Component } from "react";

interface MapProps {
    prop1: string;
}

export class MyMap<P = MapProps, E = number> extends Component<P, E> { }


let x = <Map prop1="hi" />
```

Put your cursor on `Map`

```
TypeError: Cannot read property 'flags' of undefined
    at Object.resolveAlias [as getAliasedSymbol] (C:\ls\tsserver.js:26702:36)
    at Object.getCodeActions (C:\ls\tsserver.js:85017:42)
    at C:\ls\tsserver.js:84359:33
    at Object.forEach (C:\ls\tsserver.js:1428:30)
    at Object.getFixes (C:\ls\tsserver.js:84358:16)
    at C:\ls\tsserver.js:87080:40
    at Object.forEach (C:\ls\tsserver.js:1428:30)
    at Object.getCodeFixesAtPosition (C:\ls\tsserver.js:87068:16)
    at IOSession.Session.getCodeFixes (C:\ls\tsserver.js:93345:64)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (C:\ls\tsserver.js:92338:61)
```

### LINKED ISSUE #16065: Support import code fix for JSX factory
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

See `importNameCodeFixUMDGlobalReact2.ts`.

**Expected behavior:**

Can get a code fix to add the missing import.

**Actual behavior:**

No code fix.

Warning: `const name = token.getText();` is used in a lot of places in `getImportCodeActions` (sometimes three closures deep from where it's defined!) where we should actually be using the name of the symbol we're searching for.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #15456
Does not fix #16065, but hopefully indicates how that would be fixed.

### CODE REVIEW

File: src/services/codefixes/importFixes.ts [hunk 1]
```diff
@@ -138,8 +138,22 @@ namespace ts.codefix {
...
         if (context.errorCode === Diagnostics._0_refers_to_a_UMD_global_but_the_current_file_is_a_module_Consider_adding_an_import_instead.code) {
-            const symbol = checker.getAliasedSymbol(checker.getSymbolAtLocation(token));
-            return getCodeActionForImport(symbol, /*isDefault*/ false, /*isNamespaceImport*/ true);
+            const umdSymbol = checker.getSymbolAtLocation(token);
+            let symbol: ts.Symbol;
+            let symbolName: string;
+            if (umdSymbol.flags & ts.SymbolFlags.Alias) {
+                symbol = checker.getAliasedSymbol(umdSymbol);
+                symbolName = name;
+            } else if (isJsxOpeningLikeElement(token.parent) && token.parent.tagName === token) {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/importFixes.ts]: `else` next line, turn on the linter on here please

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -2385,7 +2385,10 @@ namespace FourSlash {
             const codeFixes = this.getCodeFixActions(this.activeFile.fileName, errorCode);
 
             if (!codeFixes || codeFixes.length === 0) {
-                this.raiseError("No codefixes returned.");
+                if (expectedTextArray.length !== 0) {
+                    this.raiseError("No codefixes returned.");
+                }
+                return;
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: I don't understand why this change was made.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -210,6 +210,10 @@ namespace ts {
             getSuggestionForNonexistentProperty,
             getSuggestionForNonexistentSymbol,
             getBaseConstraintOfType,
+            getJsxNamespaceSymbol: node => {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This is a misleading name - why not `getSymbolFromJsxNamespace`?

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Also, why not just return the JSX namespace itself and look up the original entity in the `exports` for that symbol?

File: src/harness/fourslash.ts (hunk 2 above)
Comment (andy-ms) [inline src/harness/fourslash.ts]: `importNameCodeFixUMDGlobalReact2.ts` asserts that the code fixes are currently empty. Previously this method would have raised an error "No codefixes returned" even if that was exactly what you were asserting.

File: src/services/codefixes/importFixes.ts (hunk 1 above)
Comment (andy-ms) [inline src/services/codefixes/importFixes.ts]: D'oh, linter was doing nothing. #16078

File: src/compiler/checker.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: After reading through, I think I see why I was confused. JSX is *usually* a global, but in theory it could be locally imported. In that case, maybe `resolveJsxNamespaceAtLocation`

**Comment by DanielRosenwasser** [comment]
I know I've been leaving a lot of nits, but does it just make sense to expose `resolveNameAtLocation`?

**Comment by andy-ms** [comment]
@DanielRosenwasser ping
